core/types/header_ext.go

```diff
@@ -5,32 +5,210 @@ package types
 func (h *HeaderExtra) DecodeRLP(eth *ethtypes.Header, stream *rlp.Stream) error {
-	panic("not implemented")
+	temp := new(HeaderSerializable)
+	if err := stream.Decode(temp); err != nil {
+		return err
+	}
+
+	temp.updateToEth(eth)
+	temp.updateToExtras(h)
+
+	return nil
 }
 
+// EncodeJSON JSON encodes the given [ethtypes.Header] and [HeaderExtra] together
+// to the `writer`. It does merge both structs into a single [HeaderSerializable].
 func (h *HeaderExtra) EncodeJSON(eth *ethtypes.Header) ([]byte, error) {
-	panic("not implemented")
+	temp := new(HeaderSerializable)
+
+	temp.updateFromEth(eth)
+	temp.updateFromExtras(h)
+
+	return temp.EncodeJSON()
 }
 
+// DecodeJSON JSON decodes from the `input` bytes and writes the output to both the
+// [ethtypes.Header] passed as argument and to the receiver [HeaderExtra].
 func (h *HeaderExtra) DecodeJSON(eth *ethtypes.Header, input []byte) error {
-	panic("not implemented")
+	temp := new(HeaderSerializable)
+	if err := temp.DecodeJSON(input); err != nil {
+		return err
+	}
+
+	temp.updateToEth(eth)
+	temp.updateToExtras(h)
+
+	return nil
 }
 
 func (h *HeaderExtra) PostCopy(dst *ethtypes.Header) {
 	panic("not implemented")
 }
+
+func (h *HeaderSerializable) updateFromEth(eth *ethtypes.Header) {
+	h.ParentHash = eth.ParentHash
+	h.UncleHash = eth.UncleHash
+	h.Coinbase = eth.Coinbase
+	h.Root = eth.Root
+	h.TxHash = eth.TxHash
+	h.ReceiptHash = eth.ReceiptHash
+	h.Bloom = eth.Bloom
+	h.Difficulty = eth.Difficulty
+	h.Number = eth.Number
+	h.GasLimit = eth.GasLimit
+	h.GasUsed = eth.GasUsed
+	h.Time = eth.Time
+	h.Extra = eth.Extra
+	h.MixDigest = eth.MixDigest
+	h.Nonce = eth.Nonce
+	h.BaseFee = eth.BaseFee
+	h.BlobGasUsed = eth.BlobGasUsed
+	h.ExcessBlobGas = eth.ExcessBlobGas
+	h.ParentBeaconRoot = eth.ParentBeaconRoot
+}
+
+func (h *HeaderSerializable) updateToEth(eth *ethtypes.Header) {
+	eth.ParentHash = h.ParentHash
+	eth.UncleHash = h.UncleHash
+	eth.Coinbase = h.Coinbase
+	eth.Root = h.Root
+	eth.TxHash = h.TxHash
+	eth.ReceiptHash = h.ReceiptHash
+	eth.Bloom = h.Bloom
+	eth.Difficulty = h.Difficulty
+	eth.Number = h.Number
+	eth.GasLimit = h.GasLimit
+	eth.GasUsed = h.GasUsed
+	eth.Time = h.Time
+	eth.Extra = h.Extra
+	eth.MixDigest = h.MixDigest
+	eth.Nonce = h.Nonce
+	eth.BaseFee = h.BaseFee
+	eth.BlobGasUsed = h.BlobGasUsed
+	eth.ExcessBlobGas = h.ExcessBlobGas
+	eth.ParentBeaconRoot = h.ParentBeaconRoot
+}
+
+func (h *HeaderSerializable) updateFromExtras(extras *HeaderExtra) {
+	h.ExtDataHash = extras.ExtDataHash
+	h.ExtDataGasUsed = extras.ExtDataGasUsed
+	h.BlockGasCost = extras.BlockGasCost
+}
+
+func (h *HeaderSerializable) updateToExtras(extras *HeaderExtra) {
+	extras.ExtDataHash = h.ExtDataHash
+	extras.ExtDataGasUsed = h.ExtDataGasUsed
+	extras.BlockGasCost = h.BlockGasCost
+}
+
+//go:generate go run github.com/fjl/gencodec -type HeaderSerializable -field-override headerMarshaling -out gen_header_json.go
```

Comment:
I wonder if we should call the output file something else? like gen_header_serializable_{json/rlp}.go ?

Reply:
That's a good point, and it would conflict less when doing a geth sync on the coreth side

Reply:
Actually both generators don't support type aliases, I created:

- https://github.com/fjl/gencodec/pull/25 to add support for it for JSON generation
- https://github.com/ava-labs/libevm/pull/154 to add support for it for RLP generation

I did re-generate them using local modified versions of the generators for this to unblock the situation.

Reply:
The files were generated before we used imports (i.e. `Bloom = ethttypes.Bloom`) and I did not re-generate them since then. Now both generators fail due to the type aliases, but with those 2 features/fixes mentioned above, I could re-generate them fine.

Reply:
Let's open a coreth issue (and add a TODO comment at the codegen site) that notes this and continue IMO, what do you think?